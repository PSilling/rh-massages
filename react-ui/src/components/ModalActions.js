// react imports
import React, { Component } from 'react';

// util imports
import _t from '../utils/Translations.js';

/**
 * Custom component rendering action buttons for modal footers.
 */
class ModalActions extends Component {

  render() {
    return(
      <div className="pull-right">
        <button type="button" className="btn btn-primary" onClick={this.props.onProceed}
          style={{ 'marginRight': '5px' }} autoFocus={this.props.autoFocus}>
          {this.props.primaryLabel}
        </button>
        <button type="button" className="btn btn-default"
          onClick={this.props.onClose}>
          { _t.translate('Dismiss') }
        </button>
      </div>
    );
  }
}

export default ModalActions
