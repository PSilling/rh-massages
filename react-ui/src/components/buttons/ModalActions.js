// react imports
import React from "react";
import PropTypes from "prop-types";

// util imports
import _t from "../../util/Translations";

/**
 * Footer buttons for Modal dialogs. Children components can be given to add more
 * case specific elements.
 */
const ModalActions = function ModalActions(props) {
  return (
    <div className="pull-right">
      {props.children}
      {props.primaryLabel === "none" ? (
        ""
      ) : (
        <button
          type="button"
          className="btn btn-primary"
          onClick={props.onProceed}
          title={props.title}
          style={{ marginRight: "5px" }}
          autoFocus={props.autoFocus}
          disabled={props.disabled}
        >
          {props.primaryLabel}
        </button>
      )}
      <button type="button" className="btn btn-default" onClick={props.onClose}>
        {_t.translate("Dismiss")}
      </button>
    </div>
  );
};

ModalActions.propTypes = {
  /** callback function triggered on close button click */
  onClose: PropTypes.func.isRequired,
  /** callback function triggered on primary button click */
  onProceed: PropTypes.func.isRequired,
  /** whether the primary button should be automatically focused */
  autoFocus: PropTypes.bool,
  /** extra elements to be provided before the two buttons */
  children: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.node), PropTypes.node]),
  /** whether the primary button should be disabled */
  disabled: PropTypes.bool,
  /** primary button label to be displayed */
  primaryLabel: PropTypes.string,
  /** primary button title */
  title: PropTypes.string
};

ModalActions.defaultProps = {
  autoFocus: false,
  children: null,
  disabled: false,
  primaryLabel: _t.translate("Proceed"),
  title: ""
};

export default ModalActions;
