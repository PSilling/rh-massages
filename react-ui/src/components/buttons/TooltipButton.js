// react imports
import React, { Component } from "react";
import PropTypes from "prop-types";

// module imports
import { Button, Tooltip } from "reactstrap";

// util imports
import Util from "../../util/Util";

/**
 * An outline button component with an attached tooltip.
 */
class TooltipButton extends Component {
  state = { tooltipActive: false };

  tooltipTarget = Util.getTooltipTargets(1)[0];

  toggleTooltip = () => {
    this.setState(prevState => ({ tooltipActive: !prevState.tooltipActive }));
  };

  render() {
    const { disabled, label, tooltip, ...rest } = this.props;
    return (
      <span className="no-print">
        <Button {...rest} id={this.tooltipTarget} outline disabled={this.props.disabled}>
          {this.props.label}
        </Button>
        {this.props.tooltip !== "" && !this.props.disabled && (
          <Tooltip
            isOpen={this.state.tooltipActive}
            target={this.tooltipTarget}
            toggle={this.toggleTooltip}
            trigger="hover"
          >
            {this.props.tooltip}
          </Tooltip>
        )}
      </span>
    );
  }
}

TooltipButton.propTypes = {
  /** function to be called on button click */
  onClick: PropTypes.func.isRequired,
  /** whether the button should be displayed as active or not */
  active: PropTypes.bool,
  /** whether the button should be disabled */
  disabled: PropTypes.bool,
  /** button label to be displayed */
  label: PropTypes.string,
  /** tooltip shown over the button */
  tooltip: PropTypes.string
};

TooltipButton.defaultProps = {
  active: false,
  disabled: false,
  label: "Click me",
  tooltip: ""
};

export default TooltipButton;
