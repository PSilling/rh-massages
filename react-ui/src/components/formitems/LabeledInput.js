// react imports
import React, { Component } from "react";
import PropTypes from "prop-types";

// module imports
import { Col, FormGroup, Label, Input, Tooltip } from "reactstrap";

// util imports
import Util from "../../util/Util";

/**
 * A combined formgroup component with a label (can have a tooltip) and a input.
 * A column is included to set input size.
 */
class LabeledInput extends Component {
  state = { tooltipActive: false };

  tooltipTarget = Util.getTooltipTargets(1)[0];

  toggleTooltip = () => {
    this.setState(prevState => ({ tooltipActive: !prevState.tooltipActive }));
  };

  handleKeyPress = event => {
    if (event.key === "Enter") {
      this.props.onEnterPress();
    }
  };

  render() {
    const { onChange, label, value, onEnterPress, options, size, tooltip, ...rest } = this.props;
    return (
      <Col md={this.props.size}>
        <FormGroup>
          <Label id={this.tooltipTarget} for={`${this.tooltipTarget}_input`}>
            {this.props.label}
          </Label>
          <Input
            {...rest}
            id={`${this.tooltipTarget}_input`}
            value={this.props.value}
            onChange={this.props.onChange}
            onKeyPress={this.handleKeyPress}
            placeholder={this.props.label}
            list={`${this.tooltipTargets}_list`}
          />
          {this.props.options.length > 0 && (
            <datalist id={`${this.tooltipTargets}_list`}>
              {this.props.options.map(item => (
                <option key={item} value={item} />
              ))}
            </datalist>
          )}
          {this.props.tooltip !== "" && (
            <Tooltip isOpen={this.state.tooltipActive} target={this.tooltipTarget} toggle={this.toggleTooltip}>
              {this.props.tooltip}
            </Tooltip>
          )}
        </FormGroup>
      </Col>
    );
  }
}

LabeledInput.propTypes = {
  /** function called on input value change */
  onChange: PropTypes.func.isRequired,
  /** input label text */
  label: PropTypes.string.isRequired,
  /** current value of the input field */
  value: PropTypes.string.isRequired,
  /** input datalist options */
  onEnterPress: PropTypes.func,
  /** input datalist options */
  options: PropTypes.arrayOf(PropTypes.string),
  /** input column length */
  size: PropTypes.string,
  /** input label tooltip */
  tooltip: PropTypes.string
};

LabeledInput.defaultProps = {
  onEnterPress() {},
  options: [],
  size: "12",
  tooltip: ""
};

export default LabeledInput;
