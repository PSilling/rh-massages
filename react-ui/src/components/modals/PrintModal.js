// react imports
import React, { Component } from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';

// component imports
import BatchButton from '../buttons/BatchButton';
import ModalActions from '../buttons/ModalActions';

// module imports
import {ModalContainer, ModalDialog} from 'react-modal-dialog';
import moment from 'moment';
import Datetime from 'react-datetime';

// util imports
import _t from '../../util/Translations';
import Util from '../../util/Util';

/**
 * Modal dialog with printing settings.
 */
class PrintModal extends Component {

  state = {active: false, massages: [], checkedRadio: 0, filter: "", from: moment(), to: moment().add(1, "day")}

  radios = ['just today', 'this week', 'this month', 'chosen month', 'custom:']

  printMassages = () => {
    var from,
        to;
    switch (this.state.checkedRadio) {
      case 0:
        from = to = moment();
        break;
      case 1:
        from = moment().startOf("isoWeek");
        to = moment().endOf("isoWeek");
        break;
      case 2:
        from = moment().startOf("month");
        to = moment().endOf("month");
        break;
      case 3:
        from = this.props.date.clone().startOf("month");
        to = this.props.date.endOf("month");
        break;
      case 4:
        from = this.state.from;
        to = this.state.to;
        break;
      default:
        Util.notify("error", "", _t.translate('An error occured!'));
    }

    Util.get(Util.FACILITIES_URL + this.props.facilityId
      + "/massages?search=" + this.state.filter
      + "&from=" + moment(from).unix() * 1000
      + "&to=" + moment(to).unix() * 1000, (json) => {
      this.props.onPrint(json.massages);
      this.toggleModal();
      setTimeout(() => window.print(), 5);
      setTimeout(() => this.props.onPrint(null), 10);
    });
  }

  changeCheck = (id) => {
    this.setState({checkedRadio: id});
  }

  changeFilter = (event) => {
    this.setState({filter: event.target.value});
  }

  changeFrom = (date) => {
    if (typeof date === 'string'
      || date.isAfter(this.state.to)) {
      return;
    }
    this.setState({from: date});
  }

  changeTo = (date) => {
    if (typeof date === 'string'
      || date.isBefore(this.state.from)) {
      return;
    }
    this.setState({to: date});
  }

  handleModalKeyPress = (event) => {
    if (event.charCode === 13 && document.activeElement === ReactDOM.findDOMNode(this.modalDialog)) {
      this.printMassages();
    }
  }

  handleInputKeyPress = (event) => {
    if (event.key === 'Enter') {
      this.printMassages();
    }
  }

  toggleModal = () => {
    this.setState({active: !this.state.active});
  }

  render() {
    return (
      <span>
        <BatchButton label={ _t.translate('Print') } onClick={this.toggleModal} />

        {this.state.active ?
          <ModalContainer onClose={this.toggleModal}>
            <ModalDialog onClose={this.toggleModal} width="50%" style={{ 'outline': 'none' }}
              tabIndex="1" onKeyPress={this.handleModalKeyPress}
              ref={(dialog) => {
                this.modalDialog = dialog;
              }}>
              <h2>
                { _t.translate('Print settings') }
              </h2>
              <hr />

              <div className="col-md-12">
                <label>{ _t.translate('Time range') }</label>
                <div style={{ 'marginBottom': '10px'}}>
                  {this.radios.map((item, index) => (
                    <label key={index} className="radio-inline">
                      <input type="radio" onChange={() => this.changeCheck(index)}
                        onKeyPress={this.handleInputKeyPress}
                        checked={this.state.checkedRadio === index}
                      />
                      { _t.translate(item) }
                    </label>
                  ))}
                </div>
              </div>

              <div className="col-md-12">
                <div className="form-group">
                  <div className="col-md-3">
                  <Datetime id="fromInput" value={this.state.from}
                    inputProps={{disabled: this.state.checkedRadio !== 4}}
                    onChange={this.changeFrom} timeFormat={false} />
                  </div>
                   <div className="col-md-1 text-center" style={{ 'marginTop': '6px' }}>
                    <label htmlFor="toInput">{ "â€“" }</label>
                  </div>
                  <div className="col-md-3">
                  <Datetime id="toInput" value={this.state.to}
                    inputProps={{disabled: this.state.checkedRadio !== 4}}
                    onChange={this.changeTo} timeFormat={false} />
                  </div>
                </div>
              </div>

              <div className="form-group col-md-12" style={{ 'marginTop': '10px'}}>
                <label htmlFor="filterInput">{ _t.translate('Filtering') }</label>
                <input id="filterInput" value={this.state.filter} className="form-control"
                  onChange={this.changeFilter} type="text" maxLength="128"
                  onFocus={Util.moveCursorToEnd} onKeyPress={this.handleInputKeyPress}
                  placeholder={ _t.translate('Masseuse, masseur or client name') } list="masseuses"
                />
                <datalist id="masseuses">
                  {this.props.masseuses.map((item) => (
                    <option key={item} value={item} />
                  ))}
                </datalist>
              </div>

              <ModalActions
                primaryLabel={ _t.translate('Print') }
                onProceed={this.printMassages}
                onClose={this.toggleModal}
                autoFocus={true}
              />
            </ModalDialog>
          </ModalContainer> : ''
        }
      </span>
    )
  }
}

PrintModal.propTypes = {
  /** unique Massage masseuses of the given Facility */
  masseuses: PropTypes.arrayOf(PropTypes.string),
  /** ID of the selected Facility */
  facilityId: PropTypes.number,
  /** current date selected in the calendar */
  date: PropTypes.instanceOf(moment).isRequired,
  /** function called on print action */
  onPrint: PropTypes.func.isRequired
};

export default PrintModal
