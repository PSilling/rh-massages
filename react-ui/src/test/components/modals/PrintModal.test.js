// react imports
import React from "react";
import { shallow } from "enzyme";

// test imports
import moment from "moment";
import LabeledInput from "../../../components/formitems/LabeledInput";
import ModalActions from "../../../components/buttons/ModalActions";
import PrintModal from "../../../components/modals/PrintModal";
import TooltipButton from "../../../components/buttons/TooltipButton";

// test mocks
jest.mock("../../../util/Fetch");

beforeAll(() => {
  Date.now = jest.fn(() => 0);
});

afterEach(() => {
  jest.resetAllMocks();
});

test("renders inside content with correct props and functionality", () => {
  const testFunction = jest.fn();
  const testMasseuses = ["test"];
  const wrapper = shallow(
    <PrintModal masseuses={testMasseuses} facilityId={1} date={moment(0)} onPrint={testFunction} withPortal={false} />
  );
  const button = wrapper.find(TooltipButton);

  button.props().onClick();
  expect(wrapper.instance().state.active).toBe(true);

  const actions = wrapper.find(ModalActions);
  const input = wrapper.find(LabeledInput);

  expect(actions.props().onClose).toBe(button.props().onClick);
  expect(input.props().options).toBe(testMasseuses);

  actions.props().onProceed();

  expect(testFunction).not.toHaveBeenCalled();
  expect(wrapper.instance().state.active).toBe(false);
  expect(wrapper).toMatchSnapshot();
});
