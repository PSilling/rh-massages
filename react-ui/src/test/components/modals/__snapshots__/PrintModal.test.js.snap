// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders inside content with correct props and functionality 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <PrintModal
    events={
      Array [
        Object {
          "bgColor": "white",
          "massage": Object {
            "client": null,
            "date": 1970-01-01T00:00:00.000Z,
            "ending": 1970-01-01T00:00:01.000Z,
            "facility": Object {
              "id": 1,
              "name": "test",
            },
            "id": 1,
            "masseuse": Object {
              "email": "test@masseuse.org",
              "masseur": true,
              "name": "Masseuse",
              "sub": "m-sub",
              "subscribed": false,
              "surname": "Test",
            },
          },
        },
      ]
    }
    onPrint={
      [MockFunction] {
        "calls": Array [
          Array [
            Array [],
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    withPortal={false}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <TooltipIconButton
          disabled={false}
          icon="print"
          onClick={[Function]}
          size="sm"
          style={
            Object {
              "border": "0px solid transparent",
            }
          }
          tooltip="Print my massage schedule"
        />,
        false,
      ],
      "className": "no-print float-right",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "disabled": false,
          "icon": "print",
          "onClick": [Function],
          "size": "sm",
          "style": Object {
            "border": "0px solid transparent",
          },
          "tooltip": "Print my massage schedule",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      false,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <TooltipIconButton
            disabled={false}
            icon="print"
            onClick={[Function]}
            size="sm"
            style={
              Object {
                "border": "0px solid transparent",
              }
            }
            tooltip="Print my massage schedule"
          />,
          false,
        ],
        "className": "no-print float-right",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "disabled": false,
            "icon": "print",
            "onClick": [Function],
            "size": "sm",
            "style": Object {
              "border": "0px solid transparent",
            },
            "tooltip": "Print my massage schedule",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        false,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
